sdk/helper/keysutil/policy.go

```diff
@@ -2392,3 +2395,201 @@ func wrapTargetPKCS8ForImport(wrappingKey *rsa.PublicKey, preppedTargetKey []byt
+	}
+
+	// validate that the first element in the certificate chain is a leaf certificate
+	if certificateChain[0].IsCA && certificateChain[0].BasicConstraintsValid {
+		return errutil.UserError{Err: "certificate in the first element is not a valid leaf certificate"}
+	}
+
+	// validate that the last element in the certificate chain is a CA certificate
+	if len(certificateChain) > 1 && !certificateChain[len(certificateChain)-1].IsCA {
+		return errutil.UserError{Err: "certificate in the last element is not a valid CA certificate"}
+	}
+
+	// NOTE: Does this check make sense?
+
+	// validate that the certificate chain contains only one leaf certificate
+	for _, certificate := range certificateChain[1 : len(certificateChain)-1] {
+		if !certificate.IsCA {
+			return errutil.UserError{Err: "provided certificate chain contains more than one leaf certificate"}
+		}
+	}
+
+	valid, err := p.validateKeyVersionCertificateKeyMatch(keyVersion, certificateChain[0].PublicKeyAlgorithm, certificateChain[0].PublicKey)
+	if err != nil {
+		prefixedErr := fmt.Errorf("could not validate key match between leaf certificate key and key entry in transit: %w", err)
+		switch err.(type) {
+		case errutil.UserError:
+			return errutil.UserError{Err: prefixedErr.Error()}
+		default:
+			return prefixedErr
+		}
+	}
+	if !valid {
+		return errors.New("leaf certificate public key does not match the selected key version")
+	}
+
+	keyEntry, err := p.safeGetKeyEntry(keyVersion)
+	if err != nil {
+		return err
+	}
+
+	// convert the certificate chain to DER format
+	derCertificates := make([][]byte, len(certificateChain))
+	for i, certificate := range certificateChain {
+		derCertificates[i] = certificate.Raw
+	}
+
+	// set the certificate chain in the key entry
+	keyEntry.CertificateChain = derCertificates
+
+	p.Keys[strconv.Itoa(keyVersion)] = keyEntry
+	return p.Persist(ctx, storage)
+}
+
+func (p *Policy) validateKeyVersionCertificateKeyMatch(keyVersion int, certificatePublicKeyAlgorithm x509.PublicKeyAlgorithm, certificatePublicKey any) (bool, error) {
+	if !p.Type.SigningSupported() {
+		return false, errutil.UserError{Err: fmt.Sprintf("key type '%s' does not support signing", p.Type)}
+	}
+
+	var keyTypeMatches bool
+	switch p.Type {
+	case KeyType_ECDSA_P256, KeyType_ECDSA_P384, KeyType_ECDSA_P521:
+		if certificatePublicKeyAlgorithm == x509.ECDSA {
+			keyTypeMatches = true
+		}
+	case KeyType_ED25519:
+		if certificatePublicKeyAlgorithm == x509.Ed25519 {
+			keyTypeMatches = true
+		}
+	case KeyType_RSA2048, KeyType_RSA3072, KeyType_RSA4096:
+		if certificatePublicKeyAlgorithm == x509.RSA {
+			keyTypeMatches = true
+		}
+	}
+	if !keyTypeMatches {
+		return false, errutil.UserError{Err: fmt.Sprintf("provided leaf certificate public key algorithm '%s' does not match the transit key type '%s'", certificatePublicKeyAlgorithm, p.Type)}
+	}
+
+	keyEntry, err := p.safeGetKeyEntry(keyVersion)
```

Comment:
It seems weird to call this twice, tbh. I wonder if we could call it and pass it into this function from `PersistCertificateChain`?

Reply:
Yes, that might be better.